pushsync/pusher.go

```
@@ -0,0 +1,343 @@
+		pushed:   make(map[string]*pushedItem),
+		receipts: make(chan []byte),
+		ps:       ps,
+	}
+	go p.sync()
+	return p
+}
+
+// Close closes the pusher
+func (p *Pusher) Close() {
+	close(p.quit)
+}
+
+// sync starts a forever loop that pushes chunks to their neighbourhood
+// and receives receipts (statements of custody) for them.
+// chunks that are not acknowledged with a receipt are retried
+// not earlier than retryInterval after they were last pushed
+// the routine also updates counts of states on a tag in order
+// to monitor the proportion of saved, sent and synced chunks of
+// a file or collection
+func (p *Pusher) sync() {
+	var chunks <-chan chunk.Chunk
+	var unsubscribe func()
+	var syncedAddrs []storage.Address
+	var syncedItems []*pushedItem
+
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	// register handler for pssReceiptTopic on pss pubsub
+	deregister := p.ps.Register(pssReceiptTopic, false, func(msg []byte, _ *p2p.Peer) error {
+		return p.handleReceiptMsg(msg)
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+	ctx := context.Background()
+
+	var average uint64 = 1000
+	var measurements uint64
+
+	for {
+		select {
+
+		// retry interval timer triggers starting from new
+		case <-timer.C:
+			// initially timer is set to go off as well as every time we hit the end of push index
+			// so no wait for retryInterval needed to set  items synced
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// if subscribe was running, stop it
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+
+			log.Debug("set chunk status to synced, insert to db GC index")
+			syncedTags := make(map[uint32]int)
+			// set chunk status to synced, insert to db GC index
+			if err := p.store.Set(ctx, chunk.ModeSetSync, syncedAddrs...); err != nil {
+				log.Error("error setting chunks to synced", "err", err)
```

this is kind of fatal and i'd prefer to panic here rather than just continuing and risking database corruption